File: Irrganten rb/TextUI/textUI.rb

require 'io/console'
require_relative '../irrgarten/Directions'

module UI

  class TextUI

    #https://gist.github.com/acook/4190379
    def read_char
      STDIN.echo = false
      STDIN.raw!
    
      input = STDIN.getc.chr
      if input == "\e" 
        input << STDIN.read_nonblock(3) rescue nil
        input << STDIN.read_nonblock(2) rescue nil
      end
    ensure
      #STDIN.echo=true
      #STDIN.cooked!
    
      return input
    end

    def next_move
      print "Where? "
      got_input = false
      while (!got_input)
        c = read_char
        case c
          when "w"
            puts "UP ARROW"
            output = Irrgarten::Directions::UP
            got_input = true
          when "s"
            puts "DOWN ARROW"
            output = Irrgarten::Directions::DOWN
            got_input = true
          when "d"
            puts "RIGHT ARROW"
            output = Irrgarten::Directions::RIGHT
            got_input = true
          when "a"
            puts "LEFT ARROW"
            output = Irrgarten::Directions::LEFT
            got_input = true
          when "\u0003"
            puts "CONTROL-C"
            got_input = true
            exit(1)
          else
            #Error
        end
      end
      output
    end

    def show_game(get_game_state)
      puts "Laberinto\n"
      puts get_game_state.labyrinth
      puts "Jugadores\n" #imprimir el array de jugadores
      puts get_game_state.players
      puts "Monstruos\n" #imprimir el array de monstruos
      puts get_game_state.monsters
      puts "Indice del jugador actual\n"
      puts get_game_state.current_player
      if get_game_state.winner then
        puts "GANADOR"
      end
      puts get_game_state.log
      puts "\n" + "-------------     ---     -------------" + "\n"
    end
  end # class
end #module

File: irrgarten java/src/UI/TextUI.java

package UI;

import irrgarten.Directions;
import irrgarten.GameState;
import java.util.Scanner;


public class TextUI {
    
    private static Scanner in = new Scanner(System.in);
    
    private char readChar() {
        String s = in.nextLine();     
        return s.charAt(0);
    }
    

    public Directions nextMove() {
        System.out.print("Where? ");
        
        Directions direction = Directions.DOWN;
        boolean gotInput = false;
        
        while (!gotInput) {
            char c = readChar();
            switch(c) {
                case 'w':
                    System.out.print(" UP\n");
                    direction = Directions.UP;
                    gotInput = true;
                    break;
                case 's':
                    System.out.print(" DOWN\n");
                    direction = Directions.DOWN;
                    gotInput = true;
                    break;
                case 'd':
                    System.out.print("RIGHT\n");
                    direction = Directions.RIGHT;
                    gotInput = true;
                    break;
                case 'a':
                    System.out.print(" LEFT\n");
                    direction = Directions.LEFT;
                    gotInput = true;    
                    break;
            }
        }    
        return direction;
    }
    
    //mostrar en consola información
    public void showGame(GameState gameState) {  
        System.out.print("Laberinto:  " + "\n");
        System.out.print(gameState.getLabyrinthv().toString());
        System.out.print("\n" + "Jugadores:  " + "\n");
        System.out.print(gameState.getPlayers().toString());
        System.out.print("\n" + "Monstruos:  " + "\n");
        System.out.print(gameState.getMonsters().toString());
        System.out.print("\n" + "Jugador actual (índice):  ");
        System.out.print(gameState.getCurrentPlayer() + "\n");
        if (gameState.getWinner()) System.out.print("GANADOR" + "\n");
        System.out.print(gameState.getLog());
        System.out.print("\n" + "-------------     ---     -------------" + "\n");
    }
    
}
